proposals/001-ClusterRepresentation.md

```diff
@@ -0,0 +1,344 @@
+(Note that it is possible for both a filter chain to be define at both the virtual cluster and physical cluster level.  RPCs pass through both sets of filters
+in the order they are defined).
+
+An _physical cluster_ enumerates the brokers that comprise the physical cluster.
+
+#### Physical Brokers
+
+Each physical broker specifies the address of a broker and whether TLS is to be used.  It will also specify trust settings.
+
+TLS trust material may be provided at the virtual broker level or at the individual broker level.
+
+```yaml
+physicalClusters:
+- name: my-private-cluster
+  filters:
+  - type: Filter1
+    config:
+       foo: bar
+  - type: Filter2
+  tls:
+    trust:
+    ...
+  brokers:
+  - name: broker-1
+    address: broker-1-private:9092
+    ...
+    tls:
+      trust:
+      ...
+  - name: broker-2
+    address: broker-1-private:9092
+    ...
+    tls:
+      trust:
+```
+
+
+### High Level Responsibilities
+
+#### Connection Handling
+
+
+All declared endpoints will be bound.
+
+When a connection is made to an endpoint,  the system must resolve that connection to a virtual broker and hence a virtual cluster.
+
+To do this, it resolves the endpoint (and any SNI information) against the model.   This should yield exactly one broker belonging to a virtual cluster.
+It is an error otherwise and the connnection must be closed.
+
+If TLS is in use, the SSLContext can be generated from the virtual cluster definition.  This will be passed to Netty to let it complete the TLS handshake.
+
+The virtual broker and virtual cluster is used to identify the physical broker and physical cluster.
+
+The filter chains are constructed.
+
+The handler chain connects to the physical broker.
+
+#### Model Changes
+
+The system must reload chnages to the model dynammically, without dropping established connections.
+
+#### Broker Address Filter
+
+The BrokerAddressFilter must map the RPC reponses that contain the physical broker address to the virtual broker addresses that are resolveable to the client.  DescribeCluster response is an example of an RPC that needs to be mapped.
+
+In the case where Kroxylicious is being placed in front of a kafka cluster spanning a three-AZ cluster, the kroxylicious instances in the AZ won't need configuration to *connect* to brokers in the other AZs.  However, BrokerAddressFilter must be capable of rewriting the broker address for the whole cluster.
+This leads us to the conclusion that Broker Address filter need separate configuration, independent of that what could be derived from the virtual/physical cluster mapping.
+
+The BrokerAddressFilter will accept a mapping, which will map the addresses returned by the upstream broker to routable addresses that route to the virtual brokers of the cluster in question.
+
+To do this, we can use the existing filter configuration mechanism.
+
+```
+  - type: BrokerAddress
+    config:
+      brokerAddressMapping:
```

Comment:
If we add the `advertisedAddress` to the `physical broker` definition we already have the relationship established and I worry about having to generate the mappings twice in the config file. 

We could then provide the `brokerAddressMapping` as a derived property of the `virtual cluster` configuration when injected into the filters. However that implies the `virtual config` stops being a dumb value object and starts being a right domain object.

As an aside this thought was leading down the track of wondering about how to do Inversion Of Control (IOC) or Dependency Injection (DI) with filter instances. However I think making the derived address mapping a property of the `virtual cluster` config side steps that issue and leaves it in the long grass.